debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/TxLogPosition.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.connector.sqlserver;
+
+/**
+ * Defines a position of change in the transaction log. The position is defined as a combination of commit LSN
+ * and sequence number of the change in the given transaction.
+ * The sequence number is monotonically increasing in transaction but it is not guaranteed across multiple
+ * transactions so the combination is necessary to get total order.
+ *
+ * @author Jiri Pechanec
+ *
+ */
+public class TxLogPosition implements Nullable, Comparable<TxLogPosition> {
+    public static final TxLogPosition NULL = new TxLogPosition(null, null);
+    private final Lsn commitLsn;
+    private final Lsn inTxLsn;
+
+    private TxLogPosition(Lsn commitLsn, Lsn inTxLsn) {
+        this.commitLsn = commitLsn;
+        this.inTxLsn = inTxLsn;
+    }
+
+    public Lsn getCommitLsn() {
+        return commitLsn;
+    }
+
+    public Lsn getInTxLsn() {
+        return inTxLsn;
+    }
+
+    @Override
+    public String toString() {
+        return this == NULL ? "NULL" : commitLsn + "(" + inTxLsn + ")";
+    }
+
+    @Override
+    public int hashCode() {
+        final int prime = 31;
+        int result = 1;
+        result = prime * result + ((commitLsn == null) ? 0 : commitLsn.hashCode());
+        result = prime * result + ((inTxLsn == null) ? 0 : inTxLsn.hashCode());
+        return result;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj)
+            return true;
+        if (obj == null)
+            return false;
+        if (getClass() != obj.getClass())
+            return false;
+        TxLogPosition other = (TxLogPosition) obj;
+        if (commitLsn == null) {
+            if (other.commitLsn != null)
+                return false;
+        }
+        else if (!commitLsn.equals(other.commitLsn))
+            return false;
+        if (inTxLsn == null) {
+            if (other.inTxLsn != null)
+                return false;
+        }
+        else if (!inTxLsn.equals(other.inTxLsn))
+            return false;
+        return true;
+    }
+
+    @Override
+    public int compareTo(TxLogPosition o) {
```

Comment:
This one is still unclear to me; in the JIRA issue it said that events might not be totally ordered by commit LSN in the log. Here you are first comparing the commit LSN, though. How does this ensure the required ordering then?

Reply:
Commit LSN *is* ordered but not change LSN accross commits. So I first compare commit LSNs - if they are different we can decide the order if not we need to comapre the change LSN. The change LSNs with the *same* commit LSNs are ordered.